Documentation/Proposals/NNNN-third-party-issues.md

```diff
@@ -0,0 +1,108 @@
+# Improve robustness of recording Issues
+
+* Proposal: [SWT-NNNN](NNNN-third-party-issues.md)
+* Authors: [Rachel Brindle](https://github.com/younata), [Jonathan Grynspan](https://github.com/grynspan)
+* Status: **Awaiting review**
+* Bug: [apple/swift-testing#490](https://github.com/apple/swift-testing/issues/490)
+* Implementation: [apple/swift-testing#513](https://github.com/apple/swift-testing/pull/513)
+* Review: [To be added]
+
+## Introduction
+
+Integration with third party tools is important for the success of Swift Testing
+and the ecosystem as a whole. To support this, Swift Testing should offer tools
+more control over how custom issues are recorded and what is shown.
+
+## Motivation
+
+There are several third-party assertion libraries that developers are used to
+using, and Swift Testing should make it as easy as possible for the developers
+and maintainers of these third-party assertion libraries to integrate with
+Swift Testing.
+
+Swift Testing currently offers the `Issue.record` family of static methods,
+which provides an acceptable interface for one-off/custom assertions without
+using the `#expect` or `#require` macros. Unfortunately, the public
+`Issue.record` static method does not offer a way to precisely control the
+error information shown to the user - all Issues are recorded as "unconditional
+failure", which is false if the Issue actually happened as the result of an
+assertion failing.
+
+## Proposed solution
+
+We create a new public `Issue.record` static method, specifically to record and
+display only the information given to it:
+
+```swift
+extension Issue {
+  @discardableResult public static func record(
+    _ comment: Comment? = nil,
+    context toolContext: some Issue.Kind.ToolContext,
+    sourceLocation: SourceLocation = #_sourceLocation
+  ) -> Self
+}
+```
+
+This method is then used by tools to provide a comment of what happened, as well
+as information for the user to know what tool actually recorded the issue.
+
+## Detailed design
+
+`Issue.record` creates an issue of kind `Issue.Kind.recordedByTool`. This is a
+new case specifically for this API, and serves to hold on to the
+`Issue.Kind.ToolContext`. `ToolContext` is a protocol specifically to allow
+test tool authors to provide information about the tool that produced the issue,
+such as the name of the tool.
+
+When displaying the Issue to the console/IDE, only the comment and toolContext
+are used to generate the failure information:
+
+```swift
+struct MyToolContext: Issue.Kind.ToolContext {
+    let toolName = "Sample Tool"
```

Comment:
It is also (probably?) possible for us to get the module name of the module that contains the concrete type conforming to `ToolContext`—that's a little more reliable than a developer-supplied string here, but it's a symbol name rather than a plain English name, so it comes with a tradeoff.

Worth considering, even if it's just for the "alternatives considered" section.

Reply:
That's a very good point that I didn't even think about.

I added it to the alternatives considered section, because toolName (and other custom metadata, yes) might also include other metadata that would be either very difficult or impossible to determine (I used version info as an example, because I don't think there's a way for packages to infer what their version string is).